proto/substrait/validator/simple_extensions.proto

```diff
@@ -0,0 +1,356 @@
+// SPDX-License-Identifier: Apache-2.0
+syntax = "proto3";
+
+// This proto file defines a machine-readable form of simple extension YAML
+// files.
+
+package substrait.validator;
+
+import "google/protobuf/empty.proto";
+import "substrait/validator/type_system.proto";
+
+option csharp_namespace = "Substrait.Validator.Protobuf";
+option java_multiple_files = true;
+option java_package = "io.substrait.validator.proto";
+
+// Root message returned by the validator as a result of converting a simple
+// extension module to machine-readable form.
+message SimpleExtension {
+  // Metadata about the simple extension module.
+  ExtensionDefinition module = 1;
+
+  // All extensions defined in the type class namespace of this extension. All
+  // names are case-insensitively unique.
+  repeated ExtensionDefinition type_classes = 2;
+
+  // All type variation extensions. These are grouped by type class, because
+  // names need only be unique for a given type class.
+  repeated TypeVariationNamespace type_variations = 3;
+
+  // All functions defined in the function namespace of this extension. All
+  // names are case-insensitively unique.
+  repeated ExtensionDefinition functions = 4;
+
+  // Any resolved extensions defined by dependencies of this module. Note that
+  // these are not publicly exposed by this extension.
+  repeated ExtensionDefinition dependencies = 5;
+
+  // A type variation namespace scoped to a particular type class.
+  message TypeVariationNamespace {
+    // The type class that these variations are defined for.
+    DataType.Class class = 1;
+
+    // All variations defined for this type class. All names are
+    // case-insensitively unique.
+    repeated ExtensionDefinition variations = 2;
+  }
+}
+
+// Definition information for an extension.
+message ExtensionDefinition {
+  // URI of the extension module that defined the extension. URI matching is
+  // case sensitive.
+  string uri = 1;
+
+  // The set of names that may be used to case-insensitively refer to this
+  // extension within the scope of the above URI. For type classes and type
+  // variations there will only ever be one of these. For functions, the first
+  // name is always the compound name. The simple name will only be added when
```

Comment:
Since the names are different for the different variants, maybe we should just copy-paste the relevant fields instead of introducing a metadata message type.

Reply:
I would prefer using the proto definition to guide based on type instead of having to deal with these semantics when consumed, but I understand how this would blow up if applied everywhere.